llvm/tools/sycl-post-link/DeviceGlobals.cpp

```diff
@@ -0,0 +1,117 @@
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "DeviceGlobals.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Error.h"
+
+#include <algorithm>
+#include <cassert>
+
+using namespace llvm;
+
+namespace {
+
+constexpr StringRef SYCL_DEVICE_GLOBAL_SIZE_ATTR = "sycl-device-global-size";
+constexpr StringRef SYCL_UNIQUE_ID_ATTR = "sycl-unique-id";
+constexpr StringRef SYCL_DEVICE_IMAGE_SCOPE_ATTR = "device_image_scope";
+
+void AssertRelease(bool Cond, const char *Msg) {
+  if (!Cond)
+    report_fatal_error((Twine("DeviceGlobals.cpp: ") + Msg).str().c_str());
+}
+
+/// Checks whether the string that represents a boolean value ("true"/"false"
+/// or "1"/"0" actually represents \c false.
+///
+/// @param Value [in] "boolean as string" value.
+///
+/// @returns \c true if the value of \c Value represents \c false, \c false
+/// otherwise.
+bool isFalsey(StringRef Value) {
+  return Value.equals("false") || !Value.compare_numeric("0");
+}
+
+/// Checks whether the device global variable has the \c device_image_scope
+/// property. The variable has the property if the \c sycl-device-image-scope
+/// attribute is defined for the variable and the attribute value cannot be
+/// represented as \c false.
+///
+/// @param GV [in] Device Global variable.
+///
+/// @returns \c true if variable \c GV has the \c device_image_scope property,
+/// \c false otherwise.
+bool hasDeviceImageScope(const GlobalVariable &GV) {
+  return GV.hasAttribute(SYCL_DEVICE_IMAGE_SCOPE_ATTR) &&
+         !isFalsey(
+             GV.getAttribute(SYCL_DEVICE_IMAGE_SCOPE_ATTR).getValueAsString());
+}
+
+/// Returns the size (in bytes) of the underlying type \c T of the device
+/// global variable.
+///
+/// The function gets this value from the LLVM IR attribute \c
+/// sycl-device-global-size.
+///
+/// @param GV [in] Device Global variable.
+///
+/// @returns the size (int bytes) of the underlying type \c T of the
+/// device global variable represented in the LLVM IR by \c GV.
+uint32_t getUnderlyingTypeSize(const GlobalVariable &GV) {
+  assert(GV.hasAttribute(SYCL_DEVICE_GLOBAL_SIZE_ATTR) &&
+         "The device global variable must have the 'sycl-device-global-size' "
+         "attribute");
+  return static_cast<uint32_t>(std::stoul(
+      GV.getAttribute(SYCL_DEVICE_GLOBAL_SIZE_ATTR).getValueAsString().str()));
+}
+
+/// Returns the unique id for the device global variable.
+///
+/// The function gets this value from the LLVM IR attribute \c
+/// sycl-unique-id. If the attribute is not found for the variable
+/// an error should occur even in the release built.
+///
+/// @param GV [in] Device Global variable.
+///
+/// @returns the unique id of the device global variable represented
+/// in the LLVM IR by \c GV.
+StringRef getUniqueId(const GlobalVariable &GV) {
+  AssertRelease(
```

Comment:
We just filter the variable out if the `sycl-device-global-size` attribute is omitted, so we are sure the `getUnderlyingTypeSize` function will not be invoked for such variable. From the other point of view, we don't check the presence of the `sycl-unique-id` attribute before an attempt to get its value, I believe a runtime check is required here. Maybe the first `assert` in the the `getUnderlyingTypeSize` function is redundant.

Reply:
I just think that the type of assertions in both functions should be unified - they should either be both runtime checks or simple assertions, because in general, the effect of either attribute being incorrect/missing is the same as for another one - that's an undefined behavior.

Again, since the attribute we are checking here is supplied for us by DPC++ FE implementation and not by end users, I suggest that we assume that the LLVM IR is always technically correct and don't use runtime checks here.

Reply:
I've removed all the usages of `AssertRelease` and replaced them with a regular `assert`.